generator_v3.js

```diff
@@ -140,87 +147,90 @@ var generator_v3 = function(layout, difficulty, bingoList)
 					 * makes a goal 1/25 (4%) as likely as a goal with infrequency == 1.
 					 */
 					//console.log("cont = false, infrequency check failed");
-					cont = false;
-					continue;
+
+					// If we failed the RNG roll, continue the do while loop and try again
+					continue GoalGen;
 				}
 			}
-			var tagCount = [0, 0, 0, 0, 0];
 
-			for (var z=0; z < i; z++)
+			// If the current sheet already has this goal on it
+			if (currentSheet.some(r=> r.name === goalCandidate.name))
 			{
-				// Check if the goal generated is already on the sheet
-				if (currentSheet[indexes[z]].name == goalCandidate.name)
+				// Get a new goal
+				console.log(goalCandidate.name + " already on the board");
+				continue GoalGen;
+			}
+
+			// Check if the goal has any tags
+			if (goalCandidate.tags != null)
+			{
+				// foreach tag in the goal's tags
+				for (const tag of goalCandidate.tags)
 				{
-					// If it is get a new goal
-					//console.log("cont = false, same goal on sheet");
-					cont = false;
-					break;
+					// If the tag isn't in our list of tags yet, add it and set it to 0
+					if (typeof tagCount[tag.name] == 'undefined')
+					{
+						//console.log(tag.name + " not collected yet, adding");
+						tagCount[tag.name] = 0;
+					}
+					// Otherwise check if it's higher than it should be
+					else if (tagCount[tag.name] >= tag.max[difficulty - 1])
+					{
+						// If we've got too many of that tag, get a new goal
+						console.log(tag.name + " max reached with " + tagCount[tag.name] + " on the board");
+						continue GoalGen;
+					}
 				}
-				// Check if the goal generated has any anti synergy with anything already on the sheet
-				else if (typeof currentSheet[indexes[z]].antisynergy !== 'undefined' && typeof goalCandidate.antisynergy !== 'undefined'
-					&& currentSheet[indexes[z]].antisynergy.some(r=> goalCandidate.antisynergy.includes(r)))
+			}
+
+			// Check if the goal (and the goals on the sheet) has any antisynergies
+			if (typeof goalCandidate.antisynergy !== 'undefined')
+			{
+				// If it does, check to see if it's already on the sheet
+				if (antisynergys.some(r=> goalCandidate.antisynergy.includes(r)))
```

Comment:
Here, `goalCandidate.antisynergy` is a small list (every antisynergy in `goals_v4.js` has only one element), while `antisynergys` is a big list (constantly growing from zero up to 25). So it would make more sense to compare them the other way around.

Same argument applies to `goalCandidate.catalyst` and `goalCandidate.reactant`

Reply:
This would make even more sense, if `antisynergys`, `reactants`, and `catalysts` were a [`Set`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set).

Reply:
Opened #100.